target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c

```diff
@@ -0,0 +1,214 @@
+#define RB_ID_SOFTWARE_ID		0x0A
+#define RB_ID_SERIAL_NUMBER		0x0B
+#define RB_ID_MEMORY_SIZE		0x0D
+#define RB_ID_MAC_ADDRESS_COUNT		0x0E
+#define RB_ID_HW_OPTIONS		0x15
+#define RB_ID_WLAN_DATA			0x16
+#define RB_ID_BOARD_IDENTIFIER		0x17
+#define RB_ID_PRODUCT_NAME		0x21
+#define RB_ID_DEFCONF			0x26
+#define RB_ID_BOARD_REVISION		0x27
+
+static const char *rb_tlv_cell_name(u16 tag)
+{
+	switch (tag) {
+	case RB_ID_FLASH_INFO:
+		return "flash-info";
+	case RB_ID_MAC_ADDRESS_PACK:
+		return "base-mac-address";
+	case RB_ID_BOARD_PRODUCT_CODE:
+		return "board-product-code";
+	case RB_ID_BIOS_VERSION:
+		return "booter-version";
+	case RB_ID_SERIAL_NUMBER:
+		return "board-serial";
+	case RB_ID_MEMORY_SIZE:
+		return "mem-size";
+	case RB_ID_MAC_ADDRESS_COUNT:
+		return "mac-count";
+	case RB_ID_HW_OPTIONS:
+		return "hw-options";
+	case RB_ID_WLAN_DATA:
+		return "wlan-data";
+	case RB_ID_BOARD_IDENTIFIER:
+		return "board-identifier";
+	case RB_ID_PRODUCT_NAME:
+		return "product-name";
+	case RB_ID_DEFCONF:
+		return "defconf";
+	case RB_ID_BOARD_REVISION:
+		return "board-revision";
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+static int rb_tlv_mac_read_cb(void *priv, const char *id, int index,
+                              unsigned int offset, void *buf,
+                              size_t bytes)
+{
+	eth_addr_add(buf, index);
+
+	return 0;
+}
+
+static nvmem_cell_post_process_t rb_tlv_read_cb(u16 tag, u8 *buf)
+{
+	switch (tag) {
+	case RB_ID_MAC_ADDRESS_PACK:
+		return &rb_tlv_mac_read_cb;
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+static int rb_add_cells(struct device *dev, struct nvmem_device *nvmem,
+                        size_t data_len, u8 *data)
+{
+	struct nvmem_cell_info cell = {};
+	struct device_node *layout;
+	struct hard_config_tlv tlv;
+	unsigned int hdr_len = sizeof(struct hard_config_header);
+	unsigned int offset = 0;
+	int ret;
+
+	layout = of_nvmem_layout_get_container(nvmem);
+	if (!layout)
+		return -ENOENT;
+
+	while (offset < data_len) {
+		memcpy(&tlv, data + offset, sizeof(tlv));
+		if (offset + tlv.len >= data_len) {
+			dev_err(dev, "Out of bounds field (0x%x bytes at 0x%x)\n",
+				tlv.len, hdr_len + offset);
+			break;
+		}
+
+		if (tlv.tag != 0) {
+			cell.name = rb_tlv_cell_name(tlv.tag);
+			if (!cell.name)
+				goto skip;
+
+			cell.offset = hdr_len + offset + sizeof(tlv.tag) + sizeof(tlv.len);
+			/*
+			 * MikroTik stores MAC-s with length of 8 bytes,
+			 * but kernel expects it to be ETH_ALEN (6 bytes),
+			 * so we need to make sure that is the case.
+			 */
+			if (tlv.tag == RB_ID_MAC_ADDRESS_PACK)
+				cell.bytes = ETH_ALEN;
+			else
+				cell.bytes = tlv.len;
+			cell.np = of_get_child_by_name(layout, cell.name);
+			cell.read_post_process = rb_tlv_read_cb(tlv.tag, data + offset + sizeof(tlv));
+
+			ret = nvmem_add_one_cell(nvmem, &cell);
+			if (ret) {
+				of_node_put(layout);
+				return ret;
+			}
+		}
+skip:
+		offset += sizeof(tlv) + tlv.len;
```

Comment:
You seem to be partly reimplementing `routerboot_tag_find()` here and ignoring this bit:
https://github.com/openwrt/openwrt/blob/10a1e9b44afde93801b8be0a525f02d914f210e5/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c#L715-L716

while I understand that this is not a concern for your current use case, I wouldn't trust MikroTik not to break things behind your back: why not use `routerboot_tag_find()` directly and drastically simplify this code, especially as you're only interested in a single tag here?

Reply:
I would prefer not to rely on any external code and keep NVMEM layout driver self-contained.

I added alignment for the offset to 32 bits(Hopefully correctly), thanks for the comment.

Reply:
Hmm, I'm not sure I would call `routerboot_tag_find()` "external code" in this context. I appreciate your desire to keep the code self-contained but duplicating code to achieve very similar functionality is generally not ideal from a maintenance and resource-usage PoV. This is your call though :)